Add Example for using config_for

tak1n opened the pull request:
I found it in http://guides.rubyonrails.org/4_2_release_notes.html#railties-notable-changes but not in the guide for custom configuration.

tak1n committed:
Updated existing custom configuration for rails 5 where `config.x`
namespace is not needed anymore.

Added custom configuration through config_for which parses a yml file in
config folder.

pixeltrix commented on guides/source/configuring.md:
```diff
@@ -1124,6 +1124,29 @@ These configuration points are then available through the configuration object:
+  # config/application.rb
+  module MyApp
+    class Application < Rails::Application
+      config.x = Rails.application.config_for(:my_app)
```
The `x` namespacing is no longer required in Rails 5 - can you change to something specific, e.g:

``` ruby
config.payment = Rails.application.config_for(:payment)
```

rafaelfranca replied:
Isn't it?

pixeltrix commented:
@tak1n thanks for your PR - I just need to you to adjust for Rails 5 and then I'll merge it.

tak1n commented:
Forgot to commit with [skip ci] sry

tak1n commented:
Should I update the guides in the 4.2 stable branch with x namespacing example?

pixeltrix commented:
@tak1n Yes, I think so as we will be supporting 4.2 for a while as the last of the 4.x line - thanks

tak1n commented:
Okay I'll open a seperate pr, just updated the existing custom configuration for rails 5.

tak1n commented:
Also rebased it to just 1 commit, and again forgot to use skip ci sry again ^^

pixeltrix commented:
@tak1n can you edit the commit message to be more accurate - you took the one from the second commit which is a little misleading (remember the [ci skip] this time :smile:)

Also try to write the commit message in format [suggested in the contribution guide][1] where you have a short summary on the first line and then a more detailed description - it looks better in `git log` and on the GitHub website.

Thanks

[1]: http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#commit-your-changes

tak1n commented:
I will edit the commit to following:

```
Update custom configuration for rails 5 [skip ci]

Updated existing custom configuration for rails 5 and added custom configuration example via config_for yml file loading
```

Just asking before to prevent doing multiple commit edits :grinning:

pixeltrix commented:
@tak1n sorry, my bad - it looks like the change I was thinking of in de48913 got reverted in 43073b3. I need to confirm whether that change will be made now or not.

tak1n commented:
Okay no problem, I opened a PR for 4.2 #21996

woods commented:
I don't think that these examples are correct. As far as I've seen, `config_for` returns a plain old Hash:

https://github.com/rails/rails/blob/35caf46e86cdbe64a08e282ab60c2ab0160651e3/railties/lib/rails/application.rb#L226

So if you assign a config variable as described, you'll have to access it like this:

```ruby
Rails.configuration.payment['merchant_id']
```

When I try the code in this pull request, I get a NoMethodError on the hash:

```
irb(main):002:0> Rails.configuration.payment.merchant_id
NoMethodError: undefined method `merchant_id' for #<Hash:0x007f97904f3ce0>
```

If you want method-style access, as described here, then you could pass it through an OpenStruct:

```ruby
# config/application.rb
module MyApp
  class Application < Rails::Application
    config.payment = OpenStruct.new(Rails.application.config_for(:payment))
  end
end
```

*Then* you could access the values with a method call:

```ruby
Rails.configuration.payment.merchant_id
```

Is this correct? If so, what's the best technique to illustrate in the guide? The default hash-based functionality, or the much prettier method-based technique? I agree that method access is nicer, but it doesn't appear to be the default with `config_for`.

tak1n commented:
@woods  you're completely right, I forgot that I used everything with a `Hashie::Mash`.
I will update the pr and also fix the old pr https://github.com/rails/rails/pull/21996

tak1n committed:
fix doc for rails custom configuration through config_for [skip ci]

tak1n commented:
Old pr will be fixed through https://github.com/rails/rails/pull/22481

maclover7 commented:
Looks like this is already merged into 4-2-stable, so going to merge into master.
